modules/ixBidAdapter.js

```diff
@@ -466,6 +488,96 @@ function buildRequest(validBidRequests, bidderRequest, impressions, version) {
 
   return requests;
 }
+/**
+ * Calculates IX diagnostics values and packages them into
+ * an object
+ *
+ * @param {array} validBidRequests  The bid requests from prebid
+ * @return {Object} IX diag values for ad units
+ */
+function buildMultiFormatDiag(validBidRequests) {
+  var adUnitMap = {};
+  // create ad unit map and collect the required diag properties
+  for (let i = 0; i < validBidRequests.length; ++i) {
+    var bid = validBidRequests[i];
+    var trId = bid.transactionId;
+    if (!adUnitMap.hasOwnProperty(trId)) {
+      var adUnitdiags = {
+        isMultiFormat: false,
+        isNativeMedia: false,
+        isBanner: false,
+        isInstream: false,
+        isOutstream: false,
+        hasRenderer: false
+      };
+      adUnitMap[trId] = adUnitdiags;
+    }
+
+    if (utils.deepAccess(bid, 'mediaTypes')) {
+      if (Object.keys(bid.mediaTypes).length > 1) {
+        adUnitMap[trId].isMultiFormat = true;
+      }
+      if (utils.deepAccess(bid, 'mediaTypes.native')) {
+        adUnitMap[trId].isNativeMedia = true;
+      }
+      if (utils.deepAccess(bid, 'mediaTypes.banner')) {
+        adUnitMap[trId].isBanner = true;
+      }
+      if (utils.deepAccess(bid, 'mediaTypes.video.context') === 'outstream') {
+        adUnitMap[trId].isOutstream = true;
+        if (utils.deepAccess(bid, 'renderer') ||
+          utils.deepAccess(bid, 'mediaTypes.video.renderer')) {
+          adUnitMap[trId].hasRenderer = true;
+        }
+      }
+      if (utils.deepAccess(bid, 'mediaTypes.video.context') === 'instream') {
+        adUnitMap[trId].isInstream = true;
+      }
+    }
+  }
+
+  var ixdiagMF = {
+    mfu: 0,
+    bu: 0,
+    iu: 0,
+    nu: 0,
+    ou: 0,
+    allU: 0
+  };
+
+  // Iterate over adunit map and populate MF diag object
+  const adIds = Object.keys(adUnitMap);
+  for (let i = 0; i < adIds.length; ++i) {
```

Comment:
If each ad unit has a unique transaction id, increasing these counts can be included in the loop that populates `adUnitMap` instead of looping over the ad units again.